### ISSUE TITLE
codec option is not respected in this plugin.

### ISSUE THREAD

**Issue by ph** [logstash-plugins/logstash-input-jdbc#143]
The plugin doesn't make usage of the `codec` option in the configuration, some database schema are configured with a different charset than UTF-8. When this is the case it could lead logstash to blow up with a `java.nio.charset.UnsupportedCharsetException: ASCII-8BIT` when accessing some values.

This plugin should allow people to use codec to convert their charset to UTF-8.

**Comment by jordansissel** [comment]
-1 on codecs for this plugin. Codecs were intended to help plugins delimit events (like for tcp and stdin inputs) or parse a binary payload (like inputs redis, rabbitmq, etc). In the case of JDBC, the delimiting already happens (query results are in rows), and there's no "binary" payload to parse because we already have structure (columns).

I think this plugin should ensure that any strings created are converted to UTF-8. In practice, the codec does this in some plugins (stdin, etc), but for reasons above codecs aren't a fit for this plugin.

**Comment by ph** [comment]
@jordansissel Agree, but does this mean we will use another option name for it so users can specify the charset they use when converting?

 If we choose to use a new option, we need to make sure we communicate to the user that the `codec` wont be taken into consideration for this specific input. In the current state of LS, codec params is inherited from the base classes and will default to the plain codec.

**Comment by jordansissel** [comment]
Yeah I'm not sure... Maybe we need a way to provide a charset setting per column?

**Comment by ph** [comment]
or have an hybrid solution, if its a string apply the charset to all the columns, if its an hash apply the select charset for the specific column

**Comment by jordansissel** [comment]
hmm, I think that's doable, yeah, though for a configuration we don't yet support 'hash or string' for value validation

**Comment by purbon** [comment]
I think we somehow got the idea to implement an hybrid solution here, right? i'm going to experiment and try to implement it for this bug.

**Comment by purbon** [comment]
Might be we can also leverage sequel (our ORM used there) to make him take care of encoding, see http://sequel.jeremyevans.net/rdoc-plugins/classes/Sequel/Plugins/ForceEncoding.html as example.

**Comment by purbon** [comment]
Aso keep in mind one thing, the way we use the data here, all is expected to be converted in the execute_query method:

```

  def execute_query(queue)
    # update default parameters
    @parameters['sql_last_value'] = @sql_last_value
    execute_statement(@statement, @parameters) do |row|
      event = LogStash::Event.new(row)
      decorate(event)
      queue << event
    end
  end
```

what actually happened is that for 2.3.2, that conversion is broken.

**Comment by purbon** [comment]
I amend my self, it looks like is only the java event that is doing this conversion, but not the ruby event.

**Comment by colinsurprenant** [comment]
as discussed, it is not the responsibility of the `Event` object to do charset conversion.

in this case, since codecs are not used, we either need to provide an equivalent charset option to apply charset conversions if the processed data is not UTF-8 **OR** maybe JDBC provides options to query about the table-wide/column-wide configured charset?

Ultimately it is the input plugin/codec to perform correct charset validation/conversion to set UTF-8 values in the `Event`. 

Also not that the fact that this makes logstash blowup should be also corrected/addressed in a separate issue.

**Comment by ph** [comment]
> in this case, since codecs are not used, we either need to provide an equivalent charset option to apply charset conversions if the processed data is not UTF-8 OR maybe JDBC provides options to query about the table-wide/column-wide configured charset?

Good point, I remember it was possible directly in MySQL via the schema info, the jdbc object might even be able to do conversion on the fly?

**Comment by purbon** [comment]
@ph we can tell sequel to do the conversion for us, I'm writing now a PR that leveraged this but beforehand I also have to introduce Hsqldb, at my understanding derby is always utf, so we can not really test the integration of this tests.

**Comment by jsvd** [comment]
+1 the jdbc library, or whatever library is used to execute and parse query results should manage the character encoding

**Comment by colinsurprenant** [comment]
@jsvd agree but if its not possible to figure the query results charset through JDBC in some way then it will have to be a user-configurable option and the plugin will have to transcode the result values that will be set in the Event.

**Comment by colinsurprenant** [comment]
I believe this can be closed by #151 at this point.
